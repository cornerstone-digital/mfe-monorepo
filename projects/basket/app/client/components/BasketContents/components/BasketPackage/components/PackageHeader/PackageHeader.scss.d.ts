export type Styles = {
  'fade-in': string
  'fade-out': string
  'package-header': string
  'package-header-overline': string
  'package-header-underline': string
}

export type ClassNames = keyof Styles

declare const styles: Styles

export default styles
