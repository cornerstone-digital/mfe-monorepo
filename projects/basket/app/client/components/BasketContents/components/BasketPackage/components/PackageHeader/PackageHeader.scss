@import 'node_modules/@vfuk/web-core/resources/resources.scss';

.package-header {
  padding: $spacing-fixed-1 $spacing-fixed-2;
}

.package-header-underline {
  border-bottom: 1px solid $color-pale-grey;
}

.package-header-overline {
  border-top: 1px solid $color-pale-grey;
}
