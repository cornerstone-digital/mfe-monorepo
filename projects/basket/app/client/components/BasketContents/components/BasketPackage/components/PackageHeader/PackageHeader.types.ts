import { PackageWithHeaderStatus } from '../../BasketPackage.types'

export interface PackageHeaderProps {
  renderedPackage: PackageWithHeaderStatus | PackageWithHeaderStatus[]
  primaryPackage: PackageWithHeaderStatus
  isPairedWatch?: boolean
  title?: string
  subTitle?: string
  actionDisabled?: boolean
  showPackageHeaderUnderline?: boolean
}
