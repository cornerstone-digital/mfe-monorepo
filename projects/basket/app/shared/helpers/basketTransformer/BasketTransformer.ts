import getCookie from '@web-shop-core/helpers/getCookie'
import get from 'lodash/get'

import { sanitizeJson } from '@shared/helpers/transformUtils/transformUtils'
import sortByObjectProperty from '@helpers/sortByObjectProperty'
import { isUpgradeJourney } from '@helpers/typeCheck'
import { UserState } from '@pages/BasketPage/BasketPage.types'
import matchesConstCI from '@helpers/matchesConstCI'
import { BASKET_CONSTS } from '@constants'

export interface TransformedBasket extends BasketV2.Basket {
  isBusiness?: boolean
  isReviewMode?: boolean
  isUpgradeOrder?: boolean
  isBannerHidden?: boolean
  isSmallBusiness?: boolean
}

const sortBySequence = <T extends object>(a: T, b: T) => {
  return sortByObjectProperty('sequence' as keyof T, a, b)
}

class BasketTransformer {
  private basket: TransformedBasket = {}

  private getIsBannerHidden(basket: TransformedBasket): boolean {
    const { packages } = basket
    const isCombi = (bundle?: BasketV2.Bundle): boolean => {
      const offerGroup = get(bundle, 'priceDetails.merchandisingPromotions.offerGroup', '')
      return offerGroup?.toLowerCase() === 'combi offers' || offerGroup?.toLowerCase() === 'combined combi'
    }
    return packages ? packages.some(item => isCombi(item.bundle)) : false
  }

  private getIsBusiness(userState: UserState) {
    const customerSegment = getCookie('customerSegment')
    return customerSegment?.toLowerCase() === 'business' || userState?.accountSubCategory?.toLowerCase() === 'sole-trader'
  }

  private getIsSmallBusiness(userState: UserState) {
    return matchesConstCI(BASKET_CONSTS.ACCOUNT_TYPE_SMALLBUSINESS, userState.accountType)
  }

  public transformBasket(rawBasket: any, reviewMode?: boolean, userState: UserState = {}): TransformedBasket {
    const basket: TransformedBasket = sanitizeJson(rawBasket)

    basket.isReviewMode = reviewMode
    basket.isBusiness = this.getIsBusiness(userState)
    basket.isBannerHidden = this.getIsBannerHidden(basket)
    basket.isSmallBusiness = this.getIsSmallBusiness(userState)

    if (basket.packages) {
      basket.packages.sort(sortBySequence)
    }
    basket.isUpgradeOrder = isUpgradeJourney(basket.journey?.journeyType)

    this.basket = basket

    return this.basket
  }
}

export default new BasketTransformer()
