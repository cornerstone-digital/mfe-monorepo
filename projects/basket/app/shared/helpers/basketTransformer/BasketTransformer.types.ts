export interface TransformedBasket {
  basketId?: string
}
