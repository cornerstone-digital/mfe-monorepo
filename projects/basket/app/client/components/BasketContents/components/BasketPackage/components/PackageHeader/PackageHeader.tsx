import BasketItemHeader from '@components/BasketItemHeader'
import getABTestFeatureValue from '@helpers/getABTestFeatureValue'
import { useStore } from '@store'
import classnames from 'classnames'
import getBasketPackageProps from '../../helpers/getBasketPackageProps/getBasketPackageProps'
import { PackageHeaderProps } from './PackageHeader.types'
import { PackageProps } from '../../BasketPackage.types'

import './PackageHeader.scss'

const PackageHeader = (props: PackageHeaderProps) => {
  const {
    renderedPackage,
    primaryPackage,
    isPairedWatch = false,
    title = '',
    subTitle = '',
    showPackageHeaderUnderline = false,
    actionDisabled,
  } = props

  const { basketStore } = useStore()
  const { basket, pageError } = basketStore
  const { isReviewMode, isUpgradeOrder = false } = basket

  const packageId = Array.isArray(renderedPackage) ? '' : renderedPackage?.packageId || ''
  const accountCategory: string = primaryPackage?.accountCategory || ''
  const packageProps: PackageProps = Array.isArray(renderedPackage)
    ? {}
    : getBasketPackageProps(renderedPackage, accountCategory, basketStore, primaryPackage.packageId)
  const { onRemove, onUndo, headerStatus, changePackageLink } = packageProps
  const headerClassNames = classnames('package-header', {
    'package-header-underline': showPackageHeaderUnderline,
    'package-header-overline': isPairedWatch,
  })

  const isChangeButtonVisible = getABTestFeatureValue('changeLinkAB')
    ? !isReviewMode && isPairedWatch && !Array.isArray(renderedPackage)
    : !isReviewMode

  return (
    <div styleName={headerClassNames}>
      <BasketItemHeader
        title={title}
        subTitle={subTitle}
        headerStatus={headerStatus}
        showChangeButton={isChangeButtonVisible}
        changePackageLink={changePackageLink}
        onRemove={onRemove}
        onUndo={onUndo}
        isPackage={true}
        packageId={packageId}
        pageError={pageError}
        reviewMode={isReviewMode}
        isUpgrade={isUpgradeOrder}
        actionDisabled={actionDisabled}
      />
    </div>
  )
}

export default PackageHeader
