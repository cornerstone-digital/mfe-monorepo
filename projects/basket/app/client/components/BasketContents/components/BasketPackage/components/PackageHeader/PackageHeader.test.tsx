import { render } from '@testing-library/react'
import PackageHeader from './PackageHeader'
import { TransformedBasket } from '@shared/helpers/basketTransformer/BasketTransformer'

import pageContent from '@shared/config/content/BasketPageContent.json'

const basketPageContent: BasketPageContent.Basket = pageContent[0]?.basket as BasketPageContent.Basket

function mockStore(basket: Partial<TransformedBasket> = {}): any {
  return {
    basketStore: {
      basket: {
        basketId: '123',
        ...basket,
      },
      pageContent: basketPageContent,
    },
  }
}

jest.mock('@store', () => ({
  useStore: jest.fn().mockImplementation(() => mockStore()),
}))

describe('PackageHeader', () => {
  it('should render with default values', () => {
    const { getByText, asFragment } = render(<PackageHeader primaryPackage={{}} renderedPackage={{}} title={'test-title'} />)

    expect(getByText('test-title')).toBeInTheDocument()
    expect(getByText('Remove')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
