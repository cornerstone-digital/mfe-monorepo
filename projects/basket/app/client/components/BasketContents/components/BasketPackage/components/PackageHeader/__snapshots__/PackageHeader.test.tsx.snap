// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageHeader should render with default values 1`] = `
<DocumentFragment>
  <div
    stylename="package-header"
  >
    <div
      stylename="item-header"
    >
      <div
        class="item-title"
      >
        <h5
          class="heading heading-6 inherit w-bold margin-0"
        >
          test-title
        </h5>
      </div>
      <button
        data-selector="basket-header-action-testtitle"
        stylename="action-button"
      >
        <span
          class="action-text"
        >
          Remove
        </span>
        <span
          aria-hidden="true"
          class="wrapper inherit size-2 2  "
        >
          test-file-stub
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
